Building composite project with resolveJsonModule fails

<!-- 🚨 STOP 🚨 𝗦𝗧𝗢𝗣 🚨 𝑺𝑻𝑶𝑷 🚨

Half of all issues filed here are duplicates, answered in the FAQ, or not appropriate for the bug tracker. Even if you think you've found a *bug*, please read the FAQ first, especially the Common "Bugs" That Aren't Bugs section!

Please help us by doing the following steps before logging an issue:
  * Search: https://github.com/Microsoft/TypeScript/search?type=Issues
  * Read the FAQ: https://github.com/Microsoft/TypeScript/wiki/FAQ

Please fill in the *entire* template below.
-->

<!-- Please try to reproduce the issue with `typescript@next`. It may have already been fixed. -->
**TypeScript Version:**  3.0.0-rc

<!-- Search terms you tried before logging this (so others can find this issue more easily) -->
**Search Terms:** 
resolveJsonModule, composite project, TS6307

**Code**

Repro at https://github.com/strax/ts-3-json-bug-repro

**Expected behavior:**

With `composite: true` and `resolveJsonModule: true` set in `tsconfig.json`, JSON modules are resolved in the compilation like they are when the `composite` compiler option is not set.

**Actual behavior:**

Running `tsc` gives the following error:

```
error TS6307: File '/src/hello.json' is not in project file list. Projects must list all files or use an 'include' pattern.
```

In addition, trying to include `src/hello.json` in the `files` property fails with

```
error TS6054: File '/src/hello.json' has unsupported extension. The only supported extensions are '.ts', '.tsx', '.d.ts'.
```

**Playground Link:** <!-- A link to a TypeScript Playground "Share" link which demonstrates this behavior --> 

**Related Issues:** <!-- Did you find other bugs that looked similar? -->

--- a/src/compiler/program.ts
+++ b/src/compiler/program.ts
@@ -592,6 +592,7 @@ namespace ts {
         const programDiagnostics = createDiagnosticCollection();
         const currentDirectory = host.getCurrentDirectory();
         const supportedExtensions = getSupportedExtensions(options);
+        const supportedExtensionsWithJsonIfResolveJsonModule = options.resolveJsonModule ? [...supportedExtensions, Extension.Json] : undefined;
 
         // Map storing if there is emit blocking diagnostics for given input
         const hasEmitBlockingDiagnostics = createMap<boolean>();
@@ -1925,7 +1926,7 @@ namespace ts {
             refFile?: SourceFile): SourceFile | undefined {
 
             if (hasExtension(fileName)) {
-                if (!options.allowNonTsExtensions && !forEach(supportedExtensions, extension => fileExtensionIs(host.getCanonicalFileName(fileName), extension))) {
+                if (!options.allowNonTsExtensions && !forEach(supportedExtensionsWithJsonIfResolveJsonModule || supportedExtensions, extension => fileExtensionIs(host.getCanonicalFileName(fileName), extension))) {
                     if (fail) fail(Diagnostics.File_0_has_unsupported_extension_The_only_supported_extensions_are_1, fileName, "'" + supportedExtensions.join("', '") + "'");
                     return undefined;
                 }

--- a/src/testRunner/unittests/tsbuild.ts
+++ b/src/testRunner/unittests/tsbuild.ts
@@ -233,6 +233,38 @@ namespace ts {
                 assert.isBelow(fs.statSync("/src/core/index.js").mtimeMs, time(), "Upstream JS file should not have been rebuilt");
             });
         });
+
+        describe("tsbuild - with resolveJsonModule option", () => {
+            const projFs = loadProjectFromDisk("tests/projects/resolveJsonModuleAndComposite");
+            const allExpectedOutputs = ["/src/tests/dist/src/index.js", "/src/tests/dist/src/index.d.ts", "/src/tests/dist/src/hello.json"];
+
+            function verifyProjectWithResolveJsonModule(configFile: string, ...expectedDiagnosticMessages: DiagnosticMessage[]) {
+                const fs = projFs.shadow();
+                const host = new fakes.CompilerHost(fs);
+                const builder = createSolutionBuilder(host, buildHost, [configFile], { dry: false, force: false, verbose: false });
+                clearDiagnostics();
+                builder.buildAllProjects();
+                assertDiagnosticMessages(...expectedDiagnosticMessages);
+                if (!expectedDiagnosticMessages.length) {
+                    // Check for outputs. Not an exhaustive list
+                    for (const output of allExpectedOutputs) {
+                        assert(fs.existsSync(output), `Expect file ${output} to exist`);
+                    }
+                }
+            }
+
+            it("with resolveJsonModule and include only", () => {
+                verifyProjectWithResolveJsonModule("/src/tests/tsconfig_withInclude.json", Diagnostics.File_0_is_not_in_project_file_list_Projects_must_list_all_files_or_use_an_include_pattern);
+            });
+
+            it("with resolveJsonModule and files containing json file", () => {
+                verifyProjectWithResolveJsonModule("/src/tests/tsconfig_withFiles.json");
+            });
+
+            it("with resolveJsonModule and include and files", () => {
+                verifyProjectWithResolveJsonModule("/src/tests/tsconfig_withIncludeAndFiles.json");
+            });
+        });
     }
 
     export namespace OutFile {

--- a/tests/projects/resolveJsonModuleAndComposite/tests/src/hello.json
+++ b/tests/projects/resolveJsonModuleAndComposite/tests/src/hello.json
@@ -1,0 +1,3 @@
+{
+  "hello": "world"
+}

--- a/tests/projects/resolveJsonModuleAndComposite/tests/src/index.ts
+++ b/tests/projects/resolveJsonModuleAndComposite/tests/src/index.ts
@@ -1,0 +1,3 @@
+import hello from "./hello.json"
+
+export default hello.hello

--- a/tests/projects/resolveJsonModuleAndComposite/tests/tsconfig_withFiles.json
+++ b/tests/projects/resolveJsonModuleAndComposite/tests/tsconfig_withFiles.json
@@ -1,0 +1,15 @@
+{
+  "compilerOptions": {
+    "composite": true,
+    "target": "esnext",
+    "moduleResolution": "node",
+    "module": "commonjs",
+    "resolveJsonModule": true,
+    "esModuleInterop": true,
+    "allowSyntheticDefaultImports": true,
+    "outDir": "dist"
+  },
+  "files": [
+    "src/index.ts", "src/hello.json"
+  ]
+}

--- a/tests/projects/resolveJsonModuleAndComposite/tests/tsconfig_withInclude.json
+++ b/tests/projects/resolveJsonModuleAndComposite/tests/tsconfig_withInclude.json
@@ -1,0 +1,15 @@
+{
+  "compilerOptions": {
+    "composite": true,
+    "target": "esnext",
+    "moduleResolution": "node",
+    "module": "commonjs",
+    "resolveJsonModule": true,
+    "esModuleInterop": true,
+    "allowSyntheticDefaultImports": true,
+    "outDir": "dist"
+  },
+  "include": [
+    "src/**/*"
+  ]
+}

--- a/tests/projects/resolveJsonModuleAndComposite/tests/tsconfig_withIncludeAndFiles.json
+++ b/tests/projects/resolveJsonModuleAndComposite/tests/tsconfig_withIncludeAndFiles.json
@@ -1,0 +1,18 @@
+{
+  "compilerOptions": {
+    "composite": true,
+    "target": "esnext",
+    "moduleResolution": "node",
+    "module": "commonjs",
+    "resolveJsonModule": true,
+    "esModuleInterop": true,
+    "allowSyntheticDefaultImports": true,
+    "outDir": "dist"
+  },
+  "files": [
+    "src/hello.json"
+  ],
+  "include": [
+    "src/**/*"
+  ]
+}

